File: core/shared/src/BlokError.ts
import GlobalError from "./GlobalError";

/**
 * The categories every Blok node error falls into.
 *
 * Mirror of the proto `blok.runtime.v1.ErrorCategory` enum. Stored as string
 * values so JSON payloads (e.g. `GlobalError.context.json`) are human-readable.
 */
export const ErrorCategory = {
	VALIDATION: "VALIDATION",
	CONFIGURATION: "CONFIGURATION",
	DEPENDENCY: "DEPENDENCY",
	TIMEOUT: "TIMEOUT",
	PERMISSION: "PERMISSION",
	RATE_LIMIT: "RATE_LIMIT",
	NOT_FOUND: "NOT_FOUND",
	CONFLICT: "CONFLICT",
	CANCELLED: "CANCELLED",
	INTERNAL: "INTERNAL",
	PROTOCOL: "PROTOCOL",
	DATA: "DATA",
} as const;

export type ErrorCategory = (typeof ErrorCategory)[keyof typeof ErrorCategory];

/** How severe an error is. Default for thrown errors is `ERROR`. */
export const ErrorSeverity = {
	INFO: "INFO",
	WARN: "WARN",
	ERROR: "ERROR",
	FATAL: "FATAL",
} as const;

export type ErrorSeverity = (typeof ErrorSeverity)[keyof typeof ErrorSeverity];

/**
 * Default HTTP status per error category.
 *
 * Single source of truth — the runner uses these on `GlobalError.context.code`
 * and HTTP triggers use them as the response status. Override per-error via
 * `BlokErrorOpts.httpStatus`.
 */
export const DEFAULT_HTTP_STATUS: Readonly<Record<ErrorCategory, number>> = {
	VALIDATION: 400,
	CONFIGURATION: 500,
	DEPENDENCY: 502,
	TIMEOUT: 504,
	PERMISSION: 403,
	RATE_LIMIT: 429,
	NOT_FOUND: 404,
	CONFLICT: 409,
	CANCELLED: 499,
	INTERNAL: 500,
	PROTOCOL: 502,
	DATA: 422,
};

/** Default retryable hint per error category. */
export const DEFAULT_RETRYABLE: Readonly<Record<ErrorCategory, boolean>> = {
	VALIDATION: false,
	CONFIGURATION: false,
	DEPENDENCY: true,
	TIMEOUT: true,
	PERMISSION: false,
	RATE_LIMIT: true,
	NOT_FOUND: false,
	CONFLICT: false,
	CANCELLED: false,
	INTERNAL: false,
	PROTOCOL: false,
	DATA: false,
};

/**
 * The plain-data shape of a Blok node error.
 *
 * 1:1 mirror of the proto `blok.runtime.v1.NodeError` message in JSON form.
 * The runner-side gRPC codec converts between this shape and the proto type;
 * this module has no gRPC dependency.
 */
export interface NodeErrorPayload {
	/** Stable machine identifier — see `docs/error-codes.md`. */
	code: string;
	category: ErrorCategory;
	severity: ErrorSeverity;
	/** Node name that produced this error (auto-filled by SDKs). */
	node: string;
	/** SDK identifier, e.g. "blok-python3" (auto-filled). */
	sdk: string;
	sdkVersion: string;
	/** Runtime kind, e.g. "runtime.python3" (auto-filled). */
	runtimeKind: string;
	/** ISO 8601 timestamp of the error. */
	at: string;
	message: string;
	description: string;
	remediation: string;
	docUrl: string;
	/** Flattened cause chain — outermost cause first. */
	causes: NodeErrorPayload[];
	stack: string;
	/** Bounded slice of resolved inputs/state at error time. */
	contextSnapshot: unknown;
	httpStatus: number;
	retryable: boolean;
	retryAfterMs: number;
	/** Category-specific structured details (Zod issues, SQL state, etc.). */
	details: unknown;
}

/**
 * Constructor options for {@link BlokError}.
 *
 * `code`, `message`, and the implicit `category` (via the factory method) are
 * the seven required fields that, combined with auto-filled origin, make every
 * error self-describing.
 */
export interface BlokErrorOpts {
	/** Stable machine identifier (e.g. "POSTGRES_CONNECT_TIMEOUT"). */
	code: string;
	/** One-sentence human summary. */
	message: string;
	/** Multi-paragraph context: what was tried, why it failed. */
	description?: string;
	/** Suggested next step for the developer. */
	remediation?: string;
	/** Optional link to documentation explaining this code. */
	docUrl?: string;
	/** Underlying cause — any Error or BlokError. */
	cause?: Error | BlokError;
	/** Override default retryable hint. */
	retryable?: boolean;
	/** Suggested wait time before retrying. */
	retryAfterMs?: number;
	/** Category-specific structured details. */
	details?: unknown;
	/** Bounded slice of inputs/state at error time. */
	contextSnapshot?: unknown;
	/** Override the default HTTP status for this category. */
	httpStatus?: number;
	/** Override the default severity (`ERROR`). */
	severity?: ErrorSeverity;
	/** Override the auto-filled node name. */
	node?: string;
	/** Override the auto-filled SDK name. */
	sdk?: string;
	/** Override the auto-filled SDK version. */
	sdkVersion?: string;
	/** Override the auto-filled runtime kind. */
	runtimeKind?: string;
}

/**
 * Structured error type for Blok nodes. Extends {@link GlobalError} so it
 * remains fully compatible with existing `instanceof GlobalError` checks and
 * `GlobalError.context` consumers (HTTP trigger, RunTracker, Studio).
 *
 * Use the static factory methods (`BlokError.validation`,
 * `BlokError.dependency`, etc.) — direct construction is private.
 *
 * Auto-fills `name`, `stack`, and `at`. The runner enriches `node`, `sdk`,
 * `sdkVersion`, and `runtimeKind` when the error is sourced from a runtime
 * adapter; module nodes can override via {@link BlokErrorOpts}.
 *
 * @example
 *   throw BlokError.dependency({
 *     code: "POSTGRES_CONNECT_TIMEOUT",
 *     message: "Could not connect to Postgres within 5s",
 *     description: `Tried host=${host} port=${port}; timeout=${dur}ms`,
 *     remediation: "Check DATABASE_URL env var and network reachability",
 *     cause: err,
 *   });
 */
export default class BlokError extends GlobalError {
	readonly category: ErrorCategory;
	readonly severity: ErrorSeverity;
	readonly errorCode: string;
	readonly description: string;
	readonly remediation: string;
	readonly docUrl: string;
	readonly retryable: boolean;
	readonly retryAfterMs: number;
	readonly details: unknown;
	readonly contextSnapshot: unknown;
	readonly causes: ReadonlyArray<NodeErrorPayload>;
	readonly at: Date;
	readonly sdk: string;
	readonly sdkVersion: string;
	readonly runtimeKind: string;
	readonly httpStatus: number;
	readonly nodeName: string;

	private constructor(category: ErrorCategory, opts: BlokErrorOpts) {
		super(opts.message);
		Object.setPrototypeOf(this, BlokError.prototype);

		this.category = category;
		this.severity = opts.severity ?? ErrorSeverity.ERROR;
		this.errorCode = opts.code;
		this.description = opts.description ?? "";
		this.remediation = opts.remediation ?? "";
		this.docUrl = opts.docUrl ?? "";
		this.retryable = opts.retryable ?? DEFAULT_RETRYABLE[category];
		this.retryAfterMs = opts.retryAfterMs ?? 0;
		this.details = opts.details ?? null;
		this.contextSnapshot = opts.contextSnapshot ?? null;
		this.at = new Date();
		this.sdk = opts.sdk ?? "";
		this.sdkVersion = opts.sdkVersion ?? "";
		this.runtimeKind = opts.runtimeKind ?? "";
		this.nodeName = opts.node ?? "";
		this.httpStatus = opts.httpStatus ?? DEFAULT_HTTP_STATUS[category];

		this.causes = opts.cause ? Object.freeze(BlokError.flattenCauses(opts.cause)) : Object.freeze([]);

		// Populate the GlobalError.context fields so legacy consumers keep working.
		this.setCode(this.httpStatus);
		this.setName(this.nodeName);
		this.setStack(this.stack);
		this.setJson(this.toJSON() as unknown as Record<string, unknown>);
	}

	// =========================================================================
	// Factory methods (one per category)
	// =========================================================================

	static validation(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.VALIDATION, opts);
	}
	static configuration(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.CONFIGURATION, opts);
	}
	static dependency(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.DEPENDENCY, opts);
	}
	static timeout(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.TIMEOUT, opts);
	}
	static permission(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.PERMISSION, opts);
	}
	static rateLimit(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.RATE_LIMIT, opts);
	}
	static notFound(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.NOT_FOUND, opts);
	}
	static conflict(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.CONFLICT, opts);
	}
	static cancelled(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.CANCELLED, opts);
	}
	static internal(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.INTERNAL, opts);
	}
	static protocol(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.PROTOCOL, opts);
	}
	static data(opts: BlokErrorOpts): BlokError {
		return new BlokError(ErrorCategory.DATA, opts);
	}

	// =========================================================================
	// Conversion
	// =========================================================================

	/**
	 * Convert any thrown value into a `BlokError`.
	 *
	 * Used by the runner's auto-wrap layer so legacy code (`throw new
	 * Error("oops")`) still produces a structured error. Categorization is
	 * heuristic — recognizes existing `BlokError` (passthrough), `GlobalError`
	 * (preserves code/json), `Error` (wraps as `INTERNAL`), and anything else
	 * (stringified as `INTERNAL`).
	 */
	static fromUnknown(
		err: unknown,
		ctx?: { node?: string; sdk?: string; sdkVersion?: string; runtimeKind?: string },
	): BlokError {
		if (err instanceof BlokError) {
			return err;
		}

		if (err instanceof GlobalError) {
			const httpStatus = (err.context.code as number | undefined) ?? 500;
			const wrapped = new BlokError(ErrorCategory.INTERNAL, {
				code: "GLOBAL_ERROR",
				message: err.message ?? "Unknown error",
				details: err.context.json ?? null,
				httpStatus,
				node: ctx?.node ?? (err.context.name as string | undefined) ?? "",
				sdk: ctx?.sdk,
				sdkVersion: ctx?.sdkVersion,
				runtimeKind: ctx?.runtimeKind,
			});
			if (err.stack) wrapped.setStack(err.stack);
			return wrapped;
		}

		if (err instanceof Error) {
			return new BlokError(ErrorCategory.INTERNAL, {
				code: `UNCAUGHT_${err.name || "ERROR"}`.toUpperCase(),
				message: err.message || "Uncaught error",
				cause: err,
				node: ctx?.node,
				sdk: ctx?.sdk,
				sdkVersion: ctx?.sdkVersion,
				runtimeKind: ctx?.runtimeKind,
			});
		}

		return new BlokError(ErrorCategory.INTERNAL, {
			code: "UNCAUGHT_ERROR",
			message: typeof err === "string" ? err : JSON.stringify(err),
			node: ctx?.node,
			sdk: ctx?.sdk,
			sdkVersion: ctx?.sdkVersion,
			runtimeKind: ctx?.runtimeKind,
		});
	}

	/**
	 * Reconstruct a `BlokError` from a serialized {@link NodeErrorPayload}.
	 *
	 * Used by the runner's gRPC codec to convert proto `NodeError` messages
	 * received from SDKs back into TS-side errors.
	 */
	static fromJSON(payload: NodeErrorPayload): BlokError {
		const err = new BlokError(payload.category, {
			code: payload.code,
			message: payload.message,
			description: payload.description,
			remediation: payload.remediation,
			docUrl: payload.docUrl,
			retryable: payload.retryable,
			retryAfterMs: payload.retryAfterMs,
			details: payload.details,
			contextSnapshot: payload.contextSnapshot,
			httpStatus: payload.httpStatus,
			severity: payload.severity,
			node: payload.node,
			sdk: payload.sdk,
			sdkVersion: payload.sdkVersion,
			runtimeKind: payload.runtimeKind,
		});

		if (payload.stack) err.setStack(payload.stack);
		// Restore frozen causes from payload (causes are NodeErrorPayload[], not BlokError).
		(err as { causes: ReadonlyArray<NodeErrorPayload> }).causes = Object.freeze([...payload.causes]);
		return err;
	}

	/** Serialize to the canonical {@link NodeErrorPayload} shape (matches proto wire format). */
	toJSON(): NodeErrorPayload {
		return {
			code: this.errorCode,
			category: this.category,
			severity: this.severity,
			node: this.nodeName,
			sdk: this.sdk,
			sdkVersion: this.sdkVersion,
			runtimeKind: this.runtimeKind,
			at: this.at.toISOString(),
			message: this.message,
			description: this.description,
			remediation: this.remediation,
			docUrl: this.docUrl,
			causes: [...this.causes],
			stack: this.stack ?? "",
			contextSnapshot: this.contextSnapshot,
			httpStatus: this.httpStatus,
			retryable: this.retryable,
			retryAfterMs: this.retryAfterMs,
			details: this.details,
		};
	}

	// =========================================================================
	// Internal helpers
	// =========================================================================

	private static flattenCauses(cause: Error | BlokError): NodeErrorPayload[] {
		const causes: NodeErrorPayload[] = [];
		let current: Error | BlokError | undefined = cause;
		const visited = new Set<unknown>();
		while (current && !visited.has(current)) {
			visited.add(current);

			if (current instanceof BlokError) {
				// Append the BlokError's own payload (with `causes` zeroed out so we
				// don't double-count) followed by its already-flattened causes. This
				// keeps the final chain flat regardless of how deeply nested the
				// caller's BlokError-as-cause chain was.
				causes.push(BlokError.causeToPayload(current));
				causes.push(...current.causes);
				break;
			}

			causes.push(BlokError.causeToPayload(current));
			const nextCause: Error | BlokError | undefined = (current as { cause?: Error | BlokError }).cause;
			current = nextCause;
		}
		return causes;
	}

	private static causeToPayload(cause: Error | BlokError): NodeErrorPayload {
		if (cause instanceof BlokError) {
			// Strip the cause's own `causes` to avoid duplication — the caller
			// (flattenCauses) appends them separately to keep the final list flat.
			return { ...cause.toJSON(), causes: [] };
		}

		return {
			code: `UNCAUGHT_${cause.name || "ERROR"}`.toUpperCase(),
			category: ErrorCategory.INTERNAL,
			severity: ErrorSeverity.ERROR,
			node: "",
			sdk: "",
			sdkVersion: "",
			runtimeKind: "",
			at: new Date().toISOString(),
			message: cause.message || "Uncaught error",
			description: "",
			remediation: "",
			docUrl: "",
			causes: [],
			stack: cause.stack ?? "",
			contextSnapshot: null,
			httpStatus: 500,
			retryable: false,
			retryAfterMs: 0,
			details: null,
		};
	}
}

File: sdks/ruby/lib/blok/errors/blok_error.rb
# frozen_string_literal: true

require "json"
require "set"
require "time"

module Blok
  module Errors
    # Structured +BlokError+ per master plan §17 — the canonical error contract
    # every Blok node SDK populates the same way. Mirrors the TypeScript
    # +BlokError+ in +core/shared/src/BlokError.ts+, the Python +BlokError+ in
    # +sdks/python3/blok/errors/blok_error.py+, the Go +BlokError+ in
    # +sdks/go/blok_error.go+, the Rust +BlokError+ in
    # +sdks/rust/src/blok_error.rs+, the Java +BlokError+ in
    # +sdks/java/src/main/java/com/blok/blok/errors/BlokError.java+, and the
    # C# +BlokError+ in +sdks/csharp/src/Blok.Core/Errors/BlokError.cs+, so
    # node authors writing in any language see the same field shape.
    #
    # Idiomatic usage (master plan §17.5 builder pattern, Ruby kwargs flavour):
    #
    #   raise Blok::Errors::BlokError.dependency(
    #     code: "POSTGRES_CONNECT_TIMEOUT",
    #     message: "Could not connect to Postgres within 5s",
    #     description: "Tried host=#{host} port=#{port}; timeout=#{dur}ms",
    #     remediation: "Check DATABASE_URL env var and network reachability",
    #     cause: e,
    #     retryable: true,
    #     retry_after_ms: 5000,
    #     details: { "host" => host, "port" => port }
    #   )
    #
    # Inherits from +StandardError+ so handlers can +raise+ it directly. The
    # legacy +NodeError+ (5 categories) stays available for back-compat. New
    # code should prefer +BlokError+.
    class BlokError < StandardError
      # 12 canonical error categories mirroring proto +blok.runtime.v1.ErrorCategory+.
      module Category
        VALIDATION    = "VALIDATION"
        CONFIGURATION = "CONFIGURATION"
        DEPENDENCY    = "DEPENDENCY"
        TIMEOUT       = "TIMEOUT"
        PERMISSION    = "PERMISSION"
        RATE_LIMIT    = "RATE_LIMIT"
        NOT_FOUND     = "NOT_FOUND"
        CONFLICT      = "CONFLICT"
        CANCELLED     = "CANCELLED"
        INTERNAL      = "INTERNAL"
        PROTOCOL      = "PROTOCOL"
        DATA          = "DATA"

        ALL = [
          VALIDATION, CONFIGURATION, DEPENDENCY, TIMEOUT, PERMISSION, RATE_LIMIT,
          NOT_FOUND, CONFLICT, CANCELLED, INTERNAL, PROTOCOL, DATA
        ].freeze

        # Default HTTP status per category — single source of truth, matches the
        # tables in every other SDK exactly.
        DEFAULT_HTTP_STATUS = {
          VALIDATION    => 400,
          CONFIGURATION => 500,
          DEPENDENCY    => 502,
          TIMEOUT       => 504,
          PERMISSION    => 403,
          RATE_LIMIT    => 429,
          NOT_FOUND     => 404,
          CONFLICT      => 409,
          CANCELLED     => 499,
          INTERNAL      => 500,
          PROTOCOL      => 502,
          DATA          => 422
        }.freeze

        # Default retryable hint per category.
        DEFAULT_RETRYABLE = {
          DEPENDENCY => true,
          TIMEOUT    => true,
          RATE_LIMIT => true
        }.tap { |h| h.default = false }.freeze

        # Parse a wire string into a category, falling back to +INTERNAL+
        # for unknown values (matches Python/Go/Rust/Java/C# behaviour).
        def self.parse(value)
          ALL.include?(value) ? value : INTERNAL
        end
      end

      # 4 severity levels mirroring proto +blok.runtime.v1.ErrorSeverity+.
      module Severity
        INFO  = "INFO"
        WARN  = "WARN"
        ERROR = "ERROR"
        FATAL = "FATAL"

        ALL = [INFO, WARN, ERROR, FATAL].freeze

        # Parse a wire string, falling back to +ERROR+.
        def self.parse(value)
          ALL.include?(value) ? value : ERROR
        end
      end

      DEFAULT_SDK_NAME       = "blok-ruby"
      DEFAULT_RUNTIME_KIND   = "runtime.ruby"
      CONTEXT_SNAPSHOT_MAX_BYTES = 4096

      attr_accessor :category, :severity, :code, :description, :remediation, :doc_url,
                    :http_status, :retryable, :retry_after_ms, :details,
                    :context_snapshot, :causes, :stack, :at,
                    :node, :sdk, :sdk_version, :runtime_kind

      # Raw message stored separately because Ruby's +Exception#message+
      # aliases to +#to_s+, so overriding +to_s+ to format
      # +"[CATEGORY] message"+ would infinite-loop if it re-read
      # +message+. We capture the original here at construction time.
      attr_reader :raw_message

      # Construct a +BlokError+. Prefer the per-category class methods
      # (+BlokError.dependency+, +BlokError.validation+, etc.) — direct
      # construction is supported but the factories pin the category at the
      # call site.
      #
      # @param category [String] One of {Category} constants.
      # @param code [String] Stable machine identifier.
      # @param message [String] One-sentence human summary.
      # @param description [String] Multi-paragraph context.
      # @param remediation [String] Suggested next step.
      # @param doc_url [String] Link to documentation.
      # @param cause [Exception, nil] Underlying cause; flattened into +causes+.
      # @param retryable [Boolean, nil] Override the per-category default.
      # @param retry_after_ms [Integer] Suggested retry-after duration in ms.
      # @param details [Object, nil] Category-specific structured details.
      # @param context_snapshot [Object, nil] Bounded slice of inputs/state.
      # @param http_status [Integer, nil] Override per-category default status.
      # @param severity [String] Defaults to +Severity::ERROR+.
      # @param node [String] Auto-enriched by the gRPC servicer if blank.
      # @param sdk [String] Auto-enriched.
      # @param sdk_version [String] Auto-enriched.
      # @param runtime_kind [String] Auto-enriched.
      # @param at [Time] Defaults to +Time.now.utc+.
      # @param stack [String, nil] Captured stack trace.
      def initialize(
        category,
        code:, message:,
        description: "", remediation: "", doc_url: "",
        cause: nil, retryable: nil, retry_after_ms: 0,
        details: nil, context_snapshot: nil, http_status: nil,
        severity: Severity::ERROR,
        node: "", sdk: "", sdk_version: "", runtime_kind: "",
        at: nil, stack: nil
      )
        super(message)
        @raw_message      = message.to_s
        @category         = Category.parse(category)
        @severity         = Severity.parse(severity)
        @code             = code.to_s
        @description      = description.to_s
        @remediation      = remediation.to_s
        @doc_url          = doc_url.to_s
        @retryable        = retryable.nil? ? Category::DEFAULT_RETRYABLE[@category] : retryable
        @retry_after_ms   = retry_after_ms.to_i
        @details          = details
        @context_snapshot = context_snapshot
        @http_status      = http_status.nil? ? Category::DEFAULT_HTTP_STATUS[@category] : http_status.to_i
        @node             = node.to_s
        @sdk              = sdk.to_s
        @sdk_version      = sdk_version.to_s
        @runtime_kind     = runtime_kind.to_s
        @at               = at || Time.now.utc
        @stack            = stack || self.class._capture_stack
        @causes           = cause.nil? ? [] : self.class.flatten_causes(cause)
      end

      # Display format: +[CATEGORY] message+
      def to_s
        "[#{@category}] #{@raw_message}"
      end

      # Override +Exception#message+ so it returns the raw message (without
      # the +"[CATEGORY] "+ prefix). +Kernel#raise+ uses +#message+ for
      # display when no overriding +to_s+ is provided, but our +to_s+ does
      # the formatting; +#message+ should remain the unformatted string.
      def message
        @raw_message
      end

      # ====== Per-category factory shortcuts =================================

      class << self
        # Builder for a +VALIDATION+ error (default 400, non-retryable).
        def validation(**opts)    = new(Category::VALIDATION, **opts)
        # Builder for a +CONFIGURATION+ error (default 500, non-retryable).
        def configuration(**opts) = new(Category::CONFIGURATION, **opts)
        # Builder for a +DEPENDENCY+ error (default 502, retryable).
        def dependency(**opts)    = new(Category::DEPENDENCY, **opts)
        # Builder for a +TIMEOUT+ error (default 504, retryable).
        def timeout(**opts)       = new(Category::TIMEOUT, **opts)
        # Builder for a +PERMISSION+ error (default 403, non-retryable).
        def permission(**opts)    = new(Category::PERMISSION, **opts)
        # Builder for a +RATE_LIMIT+ error (default 429, retryable).
        def rate_limit(**opts)    = new(Category::RATE_LIMIT, **opts)
        # Builder for a +NOT_FOUND+ error (default 404, non-retryable).
        def not_found(**opts)     = new(Category::NOT_FOUND, **opts)
        # Builder for a +CONFLICT+ error (default 409, non-retryable).
        def conflict(**opts)      = new(Category::CONFLICT, **opts)
        # Builder for a +CANCELLED+ error (default 499, non-retryable).
        def cancelled(**opts)     = new(Category::CANCELLED, **opts)
        # Builder for an +INTERNAL+ error (default 500, non-retryable).
        def internal(**opts)      = new(Category::INTERNAL, **opts)
        # Builder for a +PROTOCOL+ error (default 502, non-retryable).
        def protocol(**opts)      = new(Category::PROTOCOL, **opts)
        # Builder for a +DATA+ error (default 422, non-retryable).
        def data(**opts)          = new(Category::DATA, **opts)
      end

      # ====== Origin auto-enrichment =========================================

      # Carrier of the auto-enrichment fields the gRPC servicer fills into a
      # handler-thrown +BlokError+ when the handler didn't set them
      # explicitly.
      Origin = Struct.new(:node, :sdk, :sdk_version, :runtime_kind, keyword_init: true) do
        # Build an +Origin+ populated with the SDK constants
        # ({DEFAULT_SDK_NAME}, {DEFAULT_RUNTIME_KIND}) and the caller-provided
        # node name + SDK version.
        def self.defaults(node:, sdk_version:)
          new(
            node: node.to_s,
            sdk: BlokError::DEFAULT_SDK_NAME,
            sdk_version: sdk_version.to_s,
            runtime_kind: BlokError::DEFAULT_RUNTIME_KIND
          )
        end
      end

      # Fill in any missing origin fields. Won't overwrite explicit values.
      # @param origin [Origin]
      # @return [self]
      def apply_origin_if_missing(origin)
        return self if origin.nil?
        @node         = origin.node         if @node.empty?
        @sdk          = origin.sdk          if @sdk.empty?
        @sdk_version  = origin.sdk_version  if @sdk_version.empty?
        @runtime_kind = origin.runtime_kind if @runtime_kind.empty?
        self
      end

      # ====== Conversion ===================================================

      # Wrap any value as a +BlokError+. Used by the runner's auto-wrap layer
      # so legacy +raise StandardError+ still produces a structured error.
      #
      # Categorization heuristic:
      # * +BlokError+ — passthrough; missing origin fields filled in.
      # * +NodeError+ (legacy) — preserves message/details/cause; category=INTERNAL.
      # * +Exception+ — wraps as INTERNAL with +code=UNCAUGHT_<TYPE>+.
      # * +Hash+ — extracts +"message"+ key, full payload preserved in details.
      # * +String+ — becomes the message.
      # * +nil+ — placeholder +"node error"+.
      # * everything else — stringified, payload preserved in details.
      #
      # @param value [Object]
      # @param origin [Origin]
      # @return [BlokError]
      def self.from_unknown(value, origin:)
        case value
        when BlokError
          value.apply_origin_if_missing(origin)
          value
        when ::Blok::Errors::NodeError
          internal_kwargs = {
            code:    "UNCAUGHT_NODEERROR",
            message: value.message || "node error",
            details: value.to_hash,
            cause:   value
          }
          new(Category::INTERNAL, **internal_kwargs).apply_origin_if_missing(origin)
        when Exception
          msg = value.message
          msg = "Uncaught error" if msg.nil? || msg.empty?
          new(
            Category::INTERNAL,
            code:    _uncaught_code(value.class),
            message: msg,
            cause:   value
          ).apply_origin_if_missing(origin)
        when nil
          new(Category::INTERNAL, code: "UNCAUGHT_ERROR", message: "node error").apply_origin_if_missing(origin)
        when String
          new(
            Category::INTERNAL,
            code:    "UNCAUGHT_ERROR",
            message: value,
            details: { "message" => value }
          ).apply_origin_if_missing(origin)
        when Hash
          msg = value["message"] || value[:message]
          message = (msg.is_a?(String) && !msg.empty?) ? msg : "node error"
          new(
            Category::INTERNAL,
            code:    "UNCAUGHT_ERROR",
            message: message,
            details: _stringify_keys(value)
          ).apply_origin_if_missing(origin)
        else
          repr = value.to_s
          new(
            Category::INTERNAL,
            code:    "UNCAUGHT_ERROR",
            message: repr,
            details: { "message" => repr }
          ).apply_origin_if_missing(origin)
        end
      end

      # Lossless serialization to a Hash matching the proto wire shape
      # (snake_case keys). Inverse of {.from_hash}.
      # @return [Hash]
      def to_hash
        {
          "code"             => @code,
          "category"         => @category,
          "severity"         => @severity,
          "node"             => @node,
          "sdk"              => @sdk,
          "sdk_version"      => @sdk_version,
          "runtime_kind"     => @runtime_kind,
          "at"               => @at.utc.iso8601(9),
          "message"          => message.to_s,
          "description"      => @description,
          "remediation"      => @remediation,
          "doc_url"          => @doc_url,
          "causes"           => @causes.map { |c| c.dup },
          "stack"            => @stack,
          "context_snapshot" => @context_snapshot,
          "http_status"      => @http_status,
          "retryable"        => @retryable,
          "retry_after_ms"   => @retry_after_ms,
          "details"          => @details
        }
      end

      # Reconstruct a +BlokError+ from a Hash. Tolerates both snake_case
      # (Ruby/Python/Go convention) and camelCase (TS payload shape) keys
      # for cross-language fixture compatibility.
      # @param raw [Hash]
      # @return [BlokError]
      def self.from_hash(raw)
        category = Category.parse(_pick(raw, "category"))
        severity = Severity.parse(_pick(raw, "severity"))
        kwargs = {
          code:             _pick(raw, "code").to_s,
          message:          _pick(raw, "message").to_s,
          description:      _pick(raw, "description").to_s,
          remediation:      _pick(raw, "remediation").to_s,
          doc_url:          _pick(raw, "doc_url", "docUrl").to_s,
          retryable:        _pick(raw, "retryable"),
          retry_after_ms:   (_pick(raw, "retry_after_ms", "retryAfterMs") || 0).to_i,
          details:          raw["details"] || raw[:details],
          context_snapshot: _pick(raw, "context_snapshot", "contextSnapshot"),
          http_status:      _pick(raw, "http_status", "httpStatus"),
          severity:         severity,
          node:             _pick(raw, "node").to_s,
          sdk:              _pick(raw, "sdk").to_s,
          sdk_version:      _pick(raw, "sdk_version", "sdkVersion").to_s,
          runtime_kind:     _pick(raw, "runtime_kind", "runtimeKind").to_s,
          stack:            _pick(raw, "stack").to_s,
          at:               _parse_at(_pick(raw, "at"))
        }
        err = new(category, **kwargs)
        causes = raw["causes"] || raw[:causes]
        if causes.is_a?(Array)
          err.causes = causes.map { |c| c.is_a?(Hash) ? _stringify_keys(c) : c }
        end
        err
      end

      # ====== Cause-chain flattening =========================================

      # Walk a Throwable's +cause+ chain and produce a flat list of payloads.
      # Mirrors the Python/Go/Rust/Java/C# implementations. Cycle-safe.
      # +BlokError+ links are lifted in directly so cross-wire serialization
      # doesn't double-count nested chains.
      # @param exc [Exception]
      # @return [Array<Hash>]
      def self.flatten_causes(exc)
        causes = []
        visited = Set.new
        current = exc
        depth = 0
        while !current.nil? && depth < 32
          break if visited.include?(current.object_id)
          visited << current.object_id
          depth += 1
          if current.is_a?(BlokError)
            payload = current.to_hash
            payload["causes"] = []
            causes << payload
            current.causes.each { |c| causes << c.dup }
            return causes
          end
          causes << _exception_to_payload(current)
          current = current.cause
        end
        causes
      end

      # ====== Internal helpers ==============================================

      # @api private
      def self._exception_to_payload(exc)
        {
          "code"             => _uncaught_code(exc.class),
          "category"         => Category::INTERNAL,
          "severity"         => Severity::ERROR,
          "node"             => "",
          "sdk"              => "",
          "sdk_version"      => "",
          "runtime_kind"     => "",
          "at"               => Time.now.utc.iso8601(9),
          "message"          => (exc.message.nil? || exc.message.empty? ? "Uncaught error" : exc.message),
          "description"      => "",
          "remediation"      => "",
          "doc_url"          => "",
          "causes"           => [],
          "stack"            => (exc.backtrace || []).join("\n"),
          "context_snapshot" => nil,
          "http_status"      => 500,
          "retryable"        => false,
          "retry_after_ms"   => 0,
          "details"          => nil
        }
      end

      # Derive an +UNCAUGHT_<TYPE>+ code from an exception class. Mirrors the
      # Python +UNCAUGHT_CONNECTIONERROR+ and Java +UNCAUGHT_IOEXCEPTION+
      # conventions: simple (unqualified) class name, alphanumerics only,
      # uppercased. The Ruby version strips the leading namespace via
      # +::+ split.
      # @api private
      def self._uncaught_code(klass)
        return "UNCAUGHT_ERROR" if klass.nil?
        simple = klass.name.to_s.split("::").last.to_s
        upper = simple.gsub(/[^A-Za-z0-9]/, "").upcase
        upper.empty? ? "UNCAUGHT_ERROR" : "UNCAUGHT_#{upper}"
      end

      # @api private
      def self._capture_stack
        # `caller` from inside `initialize` skips the BlokError init frames.
        caller(2).join("\n")
      end

      # @api private
      def self._stringify_keys(h)
        h.each_with_object({}) { |(k, v), out| out[k.to_s] = v }
      end

      # @api private
      def self._pick(raw, *keys)
        keys.each do |k|
          return raw[k] if raw.key?(k)
          sym = k.to_sym
          return raw[sym] if raw.key?(sym)
        end
        nil
      end

      # @api private
      def self._parse_at(value)
        return Time.now.utc unless value.is_a?(String) && !value.empty?
        Time.iso8601(value)
      rescue ArgumentError
        Time.now.utc
      end
    end
  end
end
